_posts/2024-02-05-java-sequenced-collections.adoc

```diff
@@ -0,0 +1,726 @@
+
+* Dans une `List`, les éléments sont parcourus dans leur ordre d’insertion et chaque élément possède un index
+* Dans une `SortedSet`, les éléments sont parcourus selon l’ordre de comparaison naturel pour des objets qui implémentent l’interface `Comparable` ou celui défini avec une implémentation de l’interface `Comparator`
+
+Il y a donc une notion de collections ordonnées et de collections triées mais aucun type dédié n’exprime cette notion dans la hiérarchie des types de l’API Collections avant le JDK 21.
+
+Une collection séquencée est un ensemble d’éléments parcourable dans un ordre défini avec un premier et un dernier élément, les éléments intermédiaires ayant un successeur et un prédécesseur.
+
+Dans le JDK 21, l’API Collections est enrichie avec les "Sequenced Collections" : la hiérarchie de types est complétée par des interfaces qui permettent d'accéder au premier et au dernier élément de la collection en utilisant les nouvelles méthodes par défaut et d'obtenir une vue inversée des éléments de la collection.
+
+== Le besoin
+
+Il n'existait pas de super-type commun pour les collections qui ont un ordre de parcours.
+
+Un autre problème est qu'il n'existe pas de méthode uniforme pour accéder au premier et au dernier élément d'une collection, ou pour parcourir ses éléments dans l'ordre inverse. 
+
+Avant Java 21, il n’est pas facile d’obtenir le premier et le dernier élément d'une collection. Exemple avec une `ArrayList` :
+
+[source,java]
+----
+package fr.sciam.java21.sequenced_collection;
+
+import java.util.List;
+
+public class MainSequencedCollection {
+
+  public static void main(String[] args) {
+
+    List<String> elements = List.of("A","B","C");
+    String premier = elements.get(0); 
+    String dernier = elements.get(elements.size() - 1);
+
+    System.out.println(premier);
+    System.out.println(dernier);
+  }
+}
+----
+
+Sans interfaces pour les définir, les opérations liées à l'ordre de parcours sont incohérentes et parfois même absentes. Bien que les implémentations permettent d'obtenir le premier ou le dernier élément, chaque collection utilise sa propre méthode, et certaines ne sont pas évidentes ou sont même absentes :
+
+[width="100%",cols="34%,33%,33%",options="header",]
+|===
+|Collection |Premier élément |Dernier élément
+|`List` |`get(0)` |`get(list.size() - 1)`
+|`Deque` |`getFirst()` |`getLast()`
+|`SortedSet` |`first()` |`last()`
+|`LinkedHashSet` |`iterator().next()` |
+|===
+
+Certaines d'entre elles sont inutilement lourdes, comme l'obtention du dernier élément d'une liste. D'autres ne sont même pas possibles sans une grande complexité, comme pour obtenir le dernier élément d'un `LinkedHashSet` qui requière d’itérer sur l'ensemble des éléments.
+
+L'itération des éléments d'une collection du premier au dernier est simple et cohérente car toutes les implémentations de Collection peuvent être parcourues vers l'avant à l'aide d'un `Iterator`.

```

Comment:
Cette phrase est longue, petite suggestion, une virgule après le mot `cohérente`

`L'itération des éléments d'une collection du premier au dernier est simple et cohérente, car toutes les implémentations de Collection peuvent être parcourues vers l'avant à l'aide d'un `Iterator`.`